pkgs/development/compilers/squeak/default.nix

```diff
@@ -1,31 +1,191 @@
 
-stdenv.mkDerivation rec {
+let
+  inherit (builtins) elemAt;
+  nullableOr = o: default: if o == null then default else o;
+
+  bits = stdenv.hostPlatform.parsed.cpu.bits;
+
+  squeakVersion = nullableOr args.squeakVersion or null "5.3-19459";
+  squeakVersionSplit = builtins.split "-" squeakVersion;
+  squeakVersionBase = elemAt squeakVersionSplit 0;
+  squeakVersionBaseSplit = lib.splitVersion squeakVersionBase;
+  squeakVersionMajor = elemAt squeakVersionBaseSplit 0;
+  squeakVersionMinor = elemAt squeakVersionBaseSplit 1;
+
+  squeakImageVersion = elemAt squeakVersionSplit 2;
+
+  squeakSourcesVersion = nullableOr args.squeakSourcesVersion or null "50";
+
+  squeakVmVersion = nullableOr args.squeakVmVersion or null "5.0-202003021730";
+  squeakVmVersionSplit = builtins.split "-" squeakVmVersion;
+  squeakVmVersionBase = elemAt squeakVmVersionSplit 0;
+  squeakVmVersionBaseSplit = lib.splitVersion squeakVmVersionBase;
+  squeakVmVersionMajor = elemAt squeakVmVersionBaseSplit 0;
+  squeakVmVersionMinor = elemAt squeakVmVersionBaseSplit 1;
+  squeakVmVersionRelease = elemAt squeakVmVersionSplit 2;
+
+  squeakVmCommitHash = nullableOr args.squeakVmCommitHash or null (fetchurl {
+    url = "https://api.github.com/repos/OpenSmalltalk/opensmalltalk-vm/commits/${squeakVmVersionRelease}";
+    curlOpts = "--header Accept:application/vnd.github.v3.sha";
+    hash = nullableOr args.squeakVmCommitHashHash or null
+      "sha256-quwmhpJlb2fp0fI9b03fBxSR44j1xmHPW20wkSqTOhQ=";
+  });
+in stdenv.mkDerivation {
   pname = "squeak";
-  version = "4.10.2.2614";
+  version = squeakVersion;
 
-  src = fetchurl {
-    sha256 = "0bpwbnpy2sb4gylchfx50sha70z36bwgdxraym4vrr93l8pd3dix";
-    url = "http://squeakvm.org/unix/release/Squeak-${version}-src.tar.gz";
+  vmVersionRelease = squeakVmVersionRelease; # "202003021730"
+  vmHash = squeakVmCommitHash;
+
+  vmSrcUrl = "https://github.com/OpenSmalltalk/opensmalltalk-vm.git";
+  src = fetchFromGitHub {
+    owner = "OpenSmalltalk";
+    repo = "opensmalltalk-vm";
+    rev = squeakVmVersion;
+    hash = nullableOr args.squeakVmHash or null
+      "sha256-rNJn5ya+7ggC21MpwSrl2ByJDjVycONKHADboH7dQLM=";
+  };
+  imageSrc = let
+    squeakImageName = "Squeak${squeakVersionBase}-${squeakImageVersion}-${toString bits}bit";
+  in fetchzip {
+    url = "https://files.squeak.org/${squeakVersionBase}/${squeakImageName}/${squeakImageName}.zip";
+    name = "source";
+    stripRoot = false;
+    hash = nullableOr args.squeakImageHash or null
+      "sha256-wDuRyc/DNqG1D4DzyBkUvrzFkBlXBtbpnANZlRV/Fas=";
+  };
+  sourcesSrc = fetchurl {
+    url = "https://files.squeak.org/sources_files/SqueakV${squeakSourcesVersion}.sources.gz";
+    hash = nullableOr args.squeakSourcesHash or null
+      "sha256-ZViZ1VgI32LwLTEyw7utp8oaAK3UmCNJnHqsGm1IKYE=";
   };
 
-  buildInputs = [ coreutils dbus freetype glib gnused libpthreadstubs
-    pango libpulseaudio which ];
-  nativeBuildInputs = [ cmake pkg-config ];
+  vmBuild = "linux64x64";
```

Comment:
Shouldn't this depend on the host architecture?

Reply:
It should. I wish all hosts shared a unified build entry point.

Reply:
Better than nothing, but `meta.platforms = [ "x86_64-linux" ]` should be set.